xcube/util/timeindex.py

```diff
@@ -0,0 +1,141 @@
+
+
+def ensure_time_label_compatible(var: Union[xr.DataArray, xr.Dataset],
+                                 labels: Dict[str, Any]) -> Dict[str, Any]:
+    """Ensure that labels['time'] is compatible with var
+
+    This function returns either the passed-in labels object, or a copy of
+    it with a modified value for labels['time'].
+
+    If there is no 'time' key in the labels dictionary or if there is no
+    'time' dimension in the var array, the original labels are returned.
+
+    If there is a 'time' key in the labels dictionary and a 'time' dimension
+    in the var array, they are checked for compatibility. If they are
+    compatible, the original labels are returned. If not, an altered labels
+    dictionary is returned, in which the time key has been modified to be
+    compatible with the type of the 'time' dimension in the var array.
+
+    See the documentation for `ensure_time_index_compatible` for details on
+    the compatibility check.
+    """
+
+    if 'time' in labels and 'time' in var.dims:
+        return dict(labels,
+                    time=ensure_time_index_compatible(var, labels['time']))
+    else:
+        return labels
+
+
+def ensure_time_index_compatible(var: Union[xr.DataArray, xr.Dataset],
+                                 timeval: Any) -> Any:
+    """Ensure that timeval is a valid time indexer for var
+
+    It is expected that the value of timeval will be a valid timestamp (i.e.
+    a valid input to pandas.Timestamp.__init__), or a slice in which the
+    start and stop fields are valid timestamps. For a slice, the start and
+    stop fields are processed separately, and their modified values (if
+    required) are returned as the start and stop fields of a new slice. The
+    step field is included unchanged in the new slice.
+
+    The compatibility check consists of checking the timezone-awareness
+    of the variable and its indexer. If the variable is timezone-aware and
+    the indexer is timezone-naive, or vice versa, a new indexer is returned
+    whose timezone-awareness corresponds to that of the variable. Otherwise
+    the original indexer is returned.
+    """
+    
+    if isinstance(timeval, slice):
+        # process start and stop separately, and pass step through unchanged
+        return slice(
+            _ensure_timestamp_compatible(var, timeval.start),
+            _ensure_timestamp_compatible(var, timeval.stop),
+            timeval.step)
+    else:
+        return _ensure_timestamp_compatible(var, timeval)
+
+
+def _ensure_timestamp_compatible(var: xr.DataArray, timeval: Any):
+    if timeval is None:
+        return None
+
+    if hasattr(timeval, "tzinfo"):
+        timestamp = timeval
+        timeval_tzinfo = timeval.tzinfo
+    else:
+        try:
+            timestamp = pd.Timestamp(timeval)
+            timeval_tzinfo = timestamp.tzinfo
+        except (TypeError, ValueError):
+            logger.warning("Can't determine indexer timezone, leaving it "
```

Comment:
https://github.com/bc-org/best-practice/blob/main/coding/python.md#quotation-marks-for-string-literals

Is there a reason for double quotes?

Reply:
Also refers to all other occurrences of logger.

Reply:
In this particular case I could claim that it saves me from having to escape the single quote within the string, but actually I just got careless :). I will go through and singularize these and any other double quotes that sneaked in.